python/pyarrow/_flight.pyx

```diff
@@ -292,12 +331,136 @@ cdef class FlightInfo:
         return result
 
 
-cdef class FlightRecordBatchReader(_CRecordBatchReader, _ReadPandasOption):
+cdef class FlightStreamReader(_CRecordBatchReader, _ReadPandasOption):
+    """A RecordBatchReader that also allows reading application metadata."""
+
     cdef dict __dict__
 
+    cdef readonly:
+        Schema schema
+
+    def read_with_metadata(self):
+        """Read the next RecordBatch along with any metadata.
+
+        Returns
+        -------
+        batch : RecordBatch
+            The next RecordBatch in the stream.
+        metadata : Buffer or None
+            Application-specific metadata for the batch as defined by
+            Flight.
+        """
+        cdef:
+            shared_ptr[CRecordBatch] batch
+            shared_ptr[CBuffer] app_metadata
+
+        with nogil:
+            check_status((<CFlightStreamReader*> self.reader.get())
+                         .ReadWithMetadata(&batch, &app_metadata))
+
+        if batch.get() == NULL:
+            raise StopIteration
+
+        if app_metadata != NULL:
+            metadata = pyarrow_wrap_buffer(app_metadata)
+        else:
+            metadata = None
+
+        return pyarrow_wrap_batch(batch), metadata
+
+
+cdef class FlightMessageReader(_CRecordBatchReader, _ReadPandasOption):
```

Comment:
So this is exactly identical to `FlightStreamReader`, except that the implementation casts its reader to `CFlightMessageReader*` instead of `CFlightStreamReader*`. Is there a way to reconcile this? Perhaps we can have a common C++ interface (with different implementations potentially)...

Reply:
Good point, I'll try to wrangle some of the duplication here.